### PR TITLE
give more information in StringIndexError

### PR DESCRIPTION
**Pull request by KristofferC**
Before

```jl
julia> str = "abcdefghκijklmno"
"abcdefghκijklmno"

julia> str[10]
ERROR: StringIndexError("abcdefghκijklmno", 10)
Stacktrace:
```

Now

```jl
julia> str = "abcdefghκijklmno"
"abcdefghκijklmno"

julia> str[10]
ERROR: StringIndexError: invalid index [10], valid previous index [9] ('κ'), valid next index [11] ('i')
```

### CODE REVIEW

**Comment by KristofferC** [comment]
Bikeshed over `valid previous index [1] ('α')`?

Perhaps `valid previous index [1]=='α'`?

File: base/strings/string.jl [hunk 1]
```diff
@@ -11,6 +11,15 @@ struct StringIndexError <: Exception
 end
 @noinline string_index_err(s::AbstractString, i::Integer) =
     throw(StringIndexError(s, Int(i)))
+function Base.showerror(io::IO, exc::StringIndexError)
+    s = exc.string
+    i = thisind(s, exc.index)
```
Comment (mbauman) [inline base/strings/string.jl:16]: `StringIndexError` is exported and may be (mis)used by user code — maybe defensively check for `exc.index < firstindex(s) && return invoke(showerror, Tuple{IO, Any}, io, exc)` before doing something that may throw?

**Comment by StefanKarpinski** [comment]
> Bikeshed over `valid previous index [1] ('α')`?

I like that version fine, it's certainly clearer to me than `[1]=='α'`.

**Comment by StefanKarpinski** [comment]
Oh but "previous valid index" reads more naturally than "valid previous index". There's some rule about how adjectives get sorted in English and "valid previous index" violates them (in some way that I only know intuitively, I have no idea what the rules actually are, just that they exist).

**Comment by StefanKarpinski** [comment]
Same for "valid next index" which would be "next valid index" instead.

**Comment by mbauman** [comment]
We use pairs to connect indices to values in some cases; it could be `nearby valid indices 9=>'κ', 11=>'i'`?  Of course the irregular plural is annoying at the end of the string, but the `previous` and `next` aren't doing much heavy lifting since I can see that 9<10<11.

**Comment by KristofferC** [comment]
> Oh but "previous valid index" reads more naturally than "valid previous index".

Yeah, I thought about this too but decided on this to sort of emphasise the "valid `nextind`" and "valid `prevind`" connection but I can change the order if it is too weird.

**Comment by KristofferC** [comment]
Did some tweaks based on comments

**Pull request merged**